Exclude buffers from dimming based on buffer-predicates

hkjels opened the pull request:
Also, renamed `dimmer-exclusion-regexp-list` to `dimmer-buffer-exclusion-regexps` for consistancy

gonewest818 commented on README.md:
```diff
@@ -54,11 +54,16 @@ Please submit pull requests with configurations for other packages!
+`dimmer-buffer-exclusion-regexps` can be used to specify buffers that
 should never be dimmed.  If the buffer name matches any regexp in this
 list then `dimmer.el` will not dim that buffer.  Note, this variable
-replaces `dimmer-exclusion-regexp` which is obsolete starting with
```
I think to avoid confusion we should keep this statement obsoleting `dimmer-exclusion-regexp` as well as the new lines you added about `dimmer-exclusion-regexp-list`.  Also in the code, see comment below.

gonewest818 commented on dimmer.el:
```diff
@@ -114,18 +114,27 @@
   :group 'dimmer)
 
 (make-obsolete-variable
- 'dimmer-exclusion-regexp
```
At this point `dimmer-exclusion-regexp` should continue to be marked obsolete because the semantics of the variable were completely different than the new regexp lists.  We should keep this code. Also, I'm pretty sure the `define-obsolete-variable-alias` macro handles marking the old variable obsolete, so it's not necessary to use `(make-obsolete-variable 'dimmer-exclustion-regexp-list ...)` if we're also providing the alias on lines 122-123.

gonewest818 commented on dimmer.el:
```diff
@@ -297,14 +306,16 @@ FRAC controls the dimming as defined in ‘dimmer-face-color’."
+                (include (cl-some (lambda (f) (funcall f buf)) dimmer-buffer-exclusion-predicates)))
+          (unless (cl-some (lambda (rxp) (string-match-p rxp name))
+                    dimmer-buffer-exclusion-regexps)
+            (when include
```
I think we can avoid nesting unless and when:
```
(unless (or (cl-some (lambda (rxp) (string-match-p rxp name))
              dimmer-buffer-exclusion-regexps)
            (cl-some (lambda (f) (funcall f buf))
              dimmer-buffer-exclusion-predicates))
   (push buf buffers)
...
```

gonewest818 reviewed (changes requested):
Thanks for putting this together! A few changes requested below.

Also it looks like this PR has some merge conflicts because I've also been committing to master, can you take a look?

hkjels commented:
These all makes sense, I'll address them when I find the time

hkjels committed:
Exclude buffers from dimming based on buffer-predicates

Fixes #25

hkjels commented:
Updated the PR

gonewest818 committed:
update documentation and fix linting errors

gonewest818 commented:
I pushed a few tweaks to get CI to pass... can you test and make sure I didn't break anything for you?

hkjels commented:
Works like a charm :)

gonewest818 committed:
fix the logic for `dimmer-buffer-exclusion-predicates`

Our implementation was wrapping the invocation of each of the
predicates `f` like `(not (funcall f buf))` which meant, when the
predicate returned nil the buffer would be excluded.  This commit
fixes the logic to match the documentation: when any predicate is
non-nil the buffer is excluded.

gonewest818 commented:
Merging now.   **Watch out** for the last commit I pushed... it "inverts" the logic for the `dimmer-buffer-exclusion-predicates` to be correct and consistent with the documentation.

With this last commit you need to change your init.el and do this:
```
(push (lambda (buf) (not (buffer-file-name buf)))
      'dimmer-buffer-exclusion-predicates)
```
